pulsar-client/src/main/java/org/apache/pulsar/client/impl/PulsarClientImpl.java

```diff
@@ -669,18 +675,105 @@ public void close() throws PulsarClientException {
     @Override
     public void shutdown() throws PulsarClientException {
         try {
-            lookup.close();
-            cnxPool.close();
-            timer.stop();
-            externalExecutorProvider.shutdownNow();
-            internalExecutorService.shutdownNow();
-            conf.getAuthentication().close();
```

Comment:
The reason for the leak is that when a close is abnormal, the following close will not be executed
So , we just need to ensure that these are closed normally. But now you shutdown the eventLoopGroup.

Reply:
what should be done then? How does close get invoked if there occurs some exception?

Reply:
@abhilashmandaliya Sorry I missed that, in some case the eventloopgroup is passed in when constructing the PulsarClient, so it's not a resource created and owned by the PulsarClient and it might be used somewhere else. So PulsarClient shouldn't just shut it down when closing itself which might break other component using that same eventloopgroup. So we should check if the eventloopgroup is created by this PulsarClient before trying to shut it down, I think a simple boolean flag should be enough.

Reply:
@MarvinCai  @315157973  
if we go with the boolean flag option, then the same needs to be passed to the ConnectionPool as it is also using the same eventLoopGroup. Putting a check on cnxPool.close at the caller site is not safe as ConnectionPool.close is closing some other resources as well. Adding such boolean parameter to the public constructor looks odd. So we can create a protected or package private constructor in ConnectionPool which will accept this new boolean param. Does that look good to you people or you have some other better choice here?

Reply:
Discussed with Abhilash, we think it is reasonable to just close whatever resources the PulsarClientImpl create and own when constructor error out, e.g. a cnxPool or eventLoopGroup & cnxPool, and user should be responsible to close whatever resources they passed in the PulsarClientImpl's constructor when catching the exception. We will update the documentation of the API to remind user.
With this change, situation should only be better in exception case since now the PulsarClientImpl constructor is not closing any resources if there's an exception.

Reply:
@MarvinCai @315157973 I have made the changes as we discussed. Also updated the java doc for the same. kindly review it and let me know if any further change is required.